iroha/src/queue.rs

```diff
@@ -1,216 +1,162 @@
+                old_tx
+                    .get_mut()
+                    .as_mut_inner_v1()
+                    .signatures
+                    .append(&mut tx.into_inner_v1().signatures);
+                return Ok(());
+            }
+            Entry::Vacant(entry) => entry,
+        };
+
+        entry.insert(tx);
+        match self.queue.push(hash) {
+            Ok(()) => Ok(()),
+            Err(hash) => Err(self.txs.remove(&hash).unwrap().1),
         }
     }
 
-    /// Gets at most `max_txs_in_block` number of transactions, but does not drop them out of the queue.
-    /// Drops only the transactions that have reached their TTL or are already in blockchain.
-    /// For MST transactions if on leader, waits for them to gather enough signatures before, showing them as output of this function.
+    /// Pops single transaction.
     ///
-    /// The reason for not dropping transaction when getting them, is that in the case of a view change this peer might become a leader,
-    /// or might need to forward transaction to the leader to check if the leader is not faulty.
-    /// If there is no view change and the block is committed then the transactions will simply drop because they are in a blockchain already.
-    #[allow(clippy::expect_used)]
-    pub fn get_pending_txs(&mut self, is_leader: bool) -> Vec<VersionedAcceptedTransaction> {
-        let mut output_txs = Vec::new();
-        let mut left_behind_txs = VecDeque::new();
-        let mut counter = self.max_txs_in_block;
-
-        while counter > 0 && !self.pending_tx_hash_queue.is_empty() {
-            let tx_hash = self
-                .pending_tx_hash_queue
-                .pop_front()
-                .expect("Unreachable, as queue not empty");
-            let tx = &self.pending_tx_by_hash[&tx_hash];
-
-            let expired = tx.is_expired(self.tx_time_to_live);
-
-            if expired {
-                iroha_logger::warn!("Transaction with hash {} dropped due to expired TTL. This can happen either due to signature condition being not satisfied in time or too many transactions in the queue.", tx_hash)
-            }
+    /// Records unsigned transaction in seen.
+    #[allow(clippy::unwrap_used, clippy::unwrap_in_result)]
+    fn pop<W: WorldTrait>(
+        &self,
+        is_leader: bool,
+        wsv: &WorldStateView<W>,
+        seen: &mut Vec<Hash>,
+    ) -> Option<VersionedAcceptedTransaction> {
+        loop {
+            let hash = self.queue.pop()?;
+            let entry = match self.txs.entry(hash) {
+                Entry::Occupied(entry) => entry,
+                Entry::Vacant(_) => unreachable!(),
+            };
 
-            if expired || tx.is_in_blockchain(&*self.wsv) {
-                self.pending_tx_by_hash
-                    .remove(&tx_hash)
-                    .expect("Should always be present, as contained in queue");
+            if entry.get().is_expired(self.ttl) {
+                iroha_logger::error!("Transaction expired");
```

Comment:
```suggestion
                iroha_logger::warn!("Transaction expired");
```
Please also add hash in the log